datafusion/core/src/physical_optimizer/sort_enforcement2.rs

```diff
@@ -0,0 +1,2872 @@
+                .map(|elem| elem.plan.clone())
+                .collect::<Vec<_>>();
+            let plan = with_new_children_if_necessary(self.plan, children_plans)?;
+            Ok(PlanWithSortRequirements {
+                plan,
+                impact_result_ordering: self.impact_result_ordering,
+                satisfy_single_distribution: self.satisfy_single_distribution,
+                required_ordering: self.required_ordering,
+                adjusted_request_ordering: self.adjusted_request_ordering,
+            })
+        }
+    }
+}
+
+impl PhysicalOptimizerRule for TopDownEnforceSorting {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        _config: &ConfigOptions,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        // Execute a Top-Down process(Preorder Traversal) to ensure the sort requirements:
+        let plan_requirements = PlanWithSortRequirements::init(plan);
+        let adjusted = plan_requirements.transform_down(&ensure_sorting)?;
+        // Execute a Top-Down process(Preorder Traversal) to remove all the unnecessary Sort
+        let adjusted_plan = adjusted.plan.transform_down(&|plan| {
+            if let Some(sort_exec) = plan.as_any().downcast_ref::<SortExec>() {
+                if ordering_satisfy(
+                    sort_exec.input().output_ordering(),
+                    sort_exec.output_ordering(),
+                    || sort_exec.input().equivalence_properties(),
+                ) {
+                    Ok(Some(Arc::new(TombStoneExec::new(
+                        sort_exec.input().clone(),
+                    ))))
+                } else {
+                    Ok(None)
+                }
+            } else {
+                Ok(None)
+            }
+        })?;
+        // Remove the TombStoneExec
+        let final_plan = adjusted_plan.transform_up(&|plan| {
+            if let Some(tombstone_exec) = plan.as_any().downcast_ref::<TombStoneExec>() {
+                Ok(Some(tombstone_exec.input.clone()))
+            } else {
+                Ok(None)
+            }
+        })?;
+        Ok(final_plan)
+    }
+
+    fn name(&self) -> &str {
+        "TopDownEnforceSorting"
+    }
+
+    fn schema_check(&self) -> bool {
+        true
+    }
+}
+
+fn ensure_sorting(
+    requirements: PlanWithSortRequirements,
+) -> Result<Option<PlanWithSortRequirements>> {
+    if let Some(sort_exec) = requirements.plan.as_any().downcast_ref::<SortExec>() {
+        // Remove unnecessary SortExec(local/global)
+        if let Some(result) = analyze_immediate_sort_removal(&requirements, sort_exec) {
+            return Ok(Some(result));
+        }
+    } else if let Some(sort_pres_exec) = requirements
+        .plan
+        .as_any()
+        .downcast_ref::<SortPreservingMergeExec>()
+    {
+        // SortPreservingMergeExec + SortExec(local/global) is the same as the global SortExec
+        // Remove unnecessary SortPreservingMergeExec + SortExec(local/global)
```

Comment:
Instead of `SortPreservingMergeExec + SortExec(local/global)`, you can use `SortExec(local/global) -> SortPreservingMergeExec`. Second reflects hierarchy better. I think.